### ISSUE TITLE
[bug] HTML5 parser loses reference to `url` parameter value

### ISSUE THREAD

**Issue by jgarber623** [sparklemotion/nokogiri#2583]
**Please describe the bug**

Invocations of `Nokogiri::HTML4()` and `Nokogiri::HTML5()` both support optional second parameters representing the parsed document's URL:

- https://github.com/sparklemotion/nokogiri/blob/main/lib/nokogiri/html4.rb#L6-L12
- https://github.com/sparklemotion/nokogiri/blob/main/lib/nokogiri/html5.rb#L25-L32

The resulting `Nokogiri::HTMLx::Document` instance responds to a `url` method that _should_ return the string value passed to `Nokogiri::HTML4`/`Nokogiri::HTML5`. This works with the HTML4 parser ([alternatively invoked](https://github.com/sparklemotion/nokogiri/blob/main/lib/nokogiri/html.rb#L7-L8) as `Nokogiri::HTML()`) but the HTML5 parser consistently returns `nil`.

I did my best to trace the code path from `Nokogiri::HTML5()` downward, but I got lost in `Nokogiri::HTML5::Document`'s private `do_parse` method ([source](https://github.com/sparklemotion/nokogiri/blob/main/lib/nokogiri/html5/document.rb#L61-L69)):

```ruby
def do_parse(string_or_io, url, encoding, options)
  string = HTML5.read_and_encode(string_or_io, encoding)
  max_attributes = options[:max_attributes] || Nokogiri::Gumbo::DEFAULT_MAX_ATTRIBUTES
  max_errors = options[:max_errors] || options[:max_parse_errors] || Nokogiri::Gumbo::DEFAULT_MAX_ERRORS
  max_depth = options[:max_tree_depth] || Nokogiri::Gumbo::DEFAULT_MAX_TREE_DEPTH
  doc = Nokogiri::Gumbo.parse(string, url, max_attributes, max_errors, max_depth, self)
  doc.encoding = "UTF-8"
  doc
end
```

Specifically, I couldn't locate source for the `Nokogiri::Gumbo.parse` call on [line 66](https://github.com/sparklemotion/nokogiri/blob/main/lib/nokogiri/html5/document.rb#L66). I suspect this code may be generated from C code which is beyond my level of understanding. Is that the case? Did I miss something?

**Help us reproduce what you're seeing**

Save the following script as `test.rb` (or similar) and run via `ruby test.rb`.

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'

  gem 'nokogiri', '~> 1.13'
end

url = 'https://hello.example'
markup = '<body>Hello, world!</body>'

puts 'Testing Nokogiri::HTML...'
puts Nokogiri::HTML(markup, url).url
# => https://hello.example

puts 'Testing Nokogiri::HTML4...'
puts Nokogiri::HTML4(markup, url).url
# => https://hello.example

puts 'Testing Nokogiri::HTML5...'
puts Nokogiri::HTML5(markup, url).url
# => nil
```

**Expected behavior**

I expected that a call to `Nokogiri::HTML5(markup, url).url` would return the value passed as the second parameter, consistent with the HTML4 parser.

**Environment**

```
# Nokogiri (1.13.6)
    ---
    warnings: []
    nokogiri:
      version: 1.13.6
      cppflags:
      - "-I/Users/jason/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/nokogiri-1.13.6-arm64-darwin/ext/nokogiri"
      - "-I/Users/jason/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/nokogiri-1.13.6-arm64-darwin/ext/nokogiri/include"
      - "-I/Users/jason/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/nokogiri-1.13.6-arm64-darwin/ext/nokogiri/include/libxml2"
      ldflags: []
    ruby:
      version: 2.7.6
      platform: arm64-darwin21
      gem_platform: arm64-darwin-21
      description: ruby 2.7.6p219 (2022-04-12 revision c9c2245c0a) [arm64-darwin21]
      engine: ruby
    libxml:
      source: packaged
      precompiled: true
      patches:
      - 0001-Remove-script-macro-support.patch
      - 0002-Update-entities-to-remove-handling-of-ssi.patch
      - 0003-libxml2.la-is-in-top_builddir.patch
      - 0004-use-glibc-strlen.patch
      - 0005-avoid-isnan-isinf.patch
      - 0006-update-automake-files-for-arm64.patch
      - '0008-htmlParseComment-handle-abruptly-closed-comments.patch'
      - '0009-allow-wildcard-namespaces.patch'
      libxml2_path: "/Users/jason/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/nokogiri-1.13.6-arm64-darwin/ext/nokogiri"
      memory_management: ruby
      iconv_enabled: true
      compiled: 2.9.14
      loaded: 2.9.14
    libxslt:
      source: packaged
      precompiled: true
      patches:
      - 0001-update-automake-files-for-arm64.patch
      datetime_enabled: true
      compiled: 1.1.35
      loaded: 1.1.35
    other_libraries:
      zlib: 1.2.12
      libiconv: '1.16'
      libgumbo: 1.0.0-nokogiri
```

**Additional context**

I'm currently using the HTML4 parser in some code and relying on the `url` parameter to resolve relative URLs in parsed markup. I discovered the above documented inconsistency while updating that code to use the newer HTML5 parser.

Thanks for reading and I appreciate your help!

**Comment by flavorjones** [comment]
@jgarber623 Thanks for reporting this! I'll take a look as soon as I can, but after a brief glance it seems like something we can easily fix.

**Comment by jgarber623** [comment]
@flavorjones Thanks for the quick fix! Tested and reviewed your changes in #2584 and everything looks good.
